### ISSUE TITLE
Jsonrpsee dependency is always compiled

### ISSUE THREAD

**Issue by DragonDev1906** [paritytech/subxt#1003]
Due to the `subxt-macro = { workspace = true }` dependency the `jsonrpsee` dependency is always included, even if the `jsonrpsee-ws` feature is not set. Under normal circumstances this is not a big issue, since `jsonrpsee` is almost always needed even if the macro is used with a path. Unfortunately, `jsonrpsee` does not compile on the `x86_64-fortanix-unknown-sgx` target, which means you cannot use subxt generated code (even if done by a build script) on that target (I only need the encoding and decoding of types, nothing related to communication or the Client).

This will also be needed for #159.

To solve this I see two options:
- Add a `macro` feature so we can use `subxt-codegen` in build.rs (build dependency) and subxt without the macro feature in the final binary (dependency)
- Add a `jsonrpsee` feature to `subxt-codegen` and `subxt-macro` that can be disabled when using subxt. This would allow us to use the macro as normally, just without fetching the metadata via url.

**Comment by jsdw** [comment]
> Add a macro feature so we can use subxt-codegen in build.rs (build dependency) and subxt without the macro feature in the final binary (dependency)

Sorry, I don't quite follow this! Subxt has a "derive" feature which can be disabled to not bring in the macro stuff, and so wouldn't this already be possible?

> which means you cannot use subxt generated code (even if done by a build script) on that target

Jsonrpsee is used in the macro invocation (via codegen) but this only happens at compile time; the actual generated code doesn't include any references to jsonrpsee whatsoever. So I think this is already ok? (and no different really than calling `subxt_codegen` in a `build.rs`? Perhaps I'm misunderstanding something though :)

**Comment by DragonDev1906** [comment]
Thank you for the quick response.

> Sorry, I don't quite follow this! Subxt has a "derive" feature which can be disabled to not bring in the macro stuff, and so wouldn't this already be possible?

It does? I couldn't find it in the [Cargo.toml](https://github.com/paritytech/subxt/blob/master/subxt/Cargo.toml), only `codec/derive` and `serde/derive` which disable macros in the two dependencies but still imports `subxt-macro`. This would be amazing to have.

> Jsonrpsee is used in the macro invocation (via subxt-codegen) but this only happens at compile time; the actual generated code doesn't include any references to jsonrpsee whatsoever. So I think this is already ok? (and no different really than calling subxt_codegen in a build.rs? Perhaps I'm misunderstanding something though :)

There is a difference when cross-compiling (unless I misunderstood how `build.rs` works): If I'm compiling on Linux for a different target (e.g. `x86_64-fortanix-unknown-sgx`), `build.rs` is compiled for `x86_64-unknown-linux-gnu`, since it has to be executed at compile time on my Linux machine (I do not know how exactly rustc does this for macros). If I list `subxt`/`subxt-codegen` only as a `[build-dependencies]` the dependencies are thus only compiled for `x86_64-unknown-linux-gnu`. When using the macro (which I'd love to do) the rust compiler seems to try compiling `subxt`/`subxt-codegen` for the target of the final binary (`x86_64-fortanix-unknown-sgx`), since it is listed as a dependency. Although jsonrpsee is indeed only used/called at compile time and will not end up in the binary, the rust compiler does not know that when compiling the dependencies, thus tries to compile it for the target which fails (compile error) because jsonrpsee is just not compatible with `x86_64-fortanix-unknown-sgx` (since compiling of dependencies happens before the tree shaking that sees that the dependencies are not needed in the final binary).

When cross-compiling a project that uses `subxt`/`subxt-macro` for `x86_64-fortanix-unknown-sgx` I don't even get to compiling the projects library crate or even compiling subxt because of this compile error in the jsonrpsee dependency.

**Comment by niklasad1** [comment]
Ok,  `subxt-macros` is not an optional dependency in subxt and forces `jsonrpsee` in dependency tree

James recall that the macros supports having a URL to remote node:

```rust
#[subxt::subxt(
     runtime_metadata_url = "wss://rpc.polkadot.io:443"
)]
mod polkadot {}
```

It won't work because `subxt-codegen` will be in your dependency tree in `sgx`. Perhaps we could feature-gate it i.e, `runtime_metadata_url` feature?

**Comment by DragonDev1906** [comment]
> Perhaps we could feature-gate it i.e, runtime_metadata_url feature?

That'd be even better, since we wouldn't need a `build.rs` at all then. I don't know if removing jsonrpsee is sufficient for getting it to run on the sgx target (it should be). Filesystem access using `std::fs` on sgx will always return an error but that shouldn't matter when executed in a macro (I think).

**Comment by DragonDev1906** [comment]
Today I learned about `proc-macro = true`. I've just looked a bit further into how exactly proc macros work and are compiled and found out that I've missed something while looking at the Cargo.toml files (although I don't know why it didn't work the first time I tried it, perhaps I'm just forgetful and didn't try this, though I'm pretty sure I did).

As it turns out @jsdw seems to be correct in that it is `no different really than calling subxt_codegen in a build.rs`. If I set `default features = false` I can use a macro because the subxt-macro dependency is indeed only build for the target the compiler is running on (kind of makes sense now that I think more about it) due to the `proc-macro = true` line.

I guess I was a bit confused because `cargo tree` returned a dependency on jsonrpsee through the subxt-macro crate, I didn't put in `default-features = false` and jumped to a conclusion too quickly.
If I run `cargo tree -e no-build,no-dev,no-proc-macro` jsonrpsee is not listed, so I guess it really doesn't matter if it is used in a build.rs file or as a proc macro. In that case the only benefit of a `runtime_metadata_url` feature would be to lower the compile times slightly.

I'm sorry for taking your time (due to me now knowing enough about proc macros).
